torch/utils/_python_dispatch.py

```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+            assert type(arg) == type(ret), f"""Called {str(func)} with input of type {type(arg)}
+and output of type {type(ret)}. But expected types to match."""
+        # Need to run under no_dispatch, because we explicitly do **not**
+        # want our subclass to intercept the set_() call.
+        # instead, our subclass should directly have its storage swapped out.
+        with torch.utils._mode_utils.no_dispatch():
+            # directly calling this overload, and passing ret.shape, because we **explicitly**
+            # don't want to reset the sizes on ret, if the storage implies a size change.
+            # Why?
+            # The purpose of this API is *not* to change the size/strides of our output- we assume it's already correct.
+            # We just want to "fix up" the storage aliasing, without modifying or output's metadata.
+            # Example: out = inp.expand(inp.shape[0], inp.shape[0])
+            #     This requires swapping the storage of out to be the same as inp,
+            #     but we do *not* want it to change the sizes/strides that were compute for out.
+            meta_in_tls = torch._C._meta_in_tls_dispatch_include()
+            # TODO: make this a context manager
+            torch._C._set_meta_in_tls_dispatch_include(True)
+            try:
+                torch.ops.aten.set_.source_Storage_storage_offset(ret, arg.untyped_storage(), 0, ret.shape)
+            finally:
+                torch._C._set_meta_in_tls_dispatch_include(meta_in_tls)
+
+    def is_read_only_alias_match(arg, ret):
+        arg_aliases = set() if not arg.alias_info else arg.alias_info.before_set
+        out_aliases = set() if not ret.alias_info else ret.alias_info.before_set
+        shared_aliases = arg_aliases & out_aliases
+        return len(shared_aliases) > 0 and not arg.alias_info.is_write
+
+    num_args = len(func._schema.arguments)
+    num_returns = len(func._schema.returns)
+    for arg_idx in range(num_args):
+        for return_idx in range(num_returns):
+            if is_read_only_alias_match(func._schema.arguments[arg_idx], func._schema.returns[return_idx]):
+                alias_storage(args[arg_idx], outs[return_idx])
+
+    # Sigh... the torchscript parser has a bug where alias annotations for Tensor[](a) don't show up properly
+    # See https://github.com/pytorch/pytorch/issues/106173
+    if func.overloadpacket in [
+        torch.ops.aten.chunk,
+        torch.ops.aten.tensor_split,
+        torch.ops.aten.split,
+        torch.ops.aten.split_with_sizes,
+        torch.ops.aten.hsplit,
+        torch.ops.aten.vsplit,
+        torch.ops.aten.dsplit,
+        torch.ops.aten.unbind,
+    ]:
+        assert isinstance(outs, list) and all(isinstance(x, torch.Tensor) for x in outs)
+        for o in outs:
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, kwargs, out):
```

Comment:
nit: not sure why "return" is in the name here?

Reply:
to make it clear that "user subclasses should return the result of this function as the output of their torch_dispatch". Open to bikeshedding if you have any suggestions.